src/wp-includes/html-api/class-wp-url.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+class WP_URL {
+	/**
+	 * Contains the original URL string.
+	 *
+	 * Example:
+	 *
+	 *     - http://wordpress.org
+	 *     - https://wordpress.org/absolute/path
+	 *     - javascript:alert(1);
+	 *     - /relative/path
+	 *     - /?with=query&string
+	 *     - /#with-fragment
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var string
+	 */
+	private $raw_url;
+
+	/**
+	 * For relative URLs, this provides the base.
+	 *
+	 * @var string|null
+	 */
+	private $base_url;
+
+	/**
+	 * Indicates if the URL is an absolute or relative reference.
+	 *
+	 * @var bool
+	 */
+	private $is_relative;
+
+	private $parser_state = self::STATE_READY;
+
+	private $scheme;
+
+	private $username;
+
+	private $password;
+
+	private $host;
+
+	private $port;
+
+	private $path;
+
+	private $query_params;
+
+	private $fragment;
+
+	/**
+	 * Creates a new instance from a potential string URL.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param string $raw_url Possibly represents a URL.
+	 * @return WP_URL Newly-created URL.
+	 */
+	public static function parse( $raw_url, $base_url = null ) {
+		if ( ! is_string( $raw_url ) || strlen( $raw_url ) <= 0 ) {
+			return null;
+		}
+
+		$url          = new WP_URL( $base_url );
+		$url->raw_url = $raw_url;
+		$at           = 0;
+
+		$first_char  = $raw_url[0];
+		$is_relative = '/' === $first_char || '?' === $first_char || '#' === $first_char;
```

Comment:
Or `./`. Technically `plugins.php` or any other string could also be a relative url, but that gets tricky

Reply:
can it? is that a relative URL if it starts with something that isn't `/`?

either way I think it's a relatively unimportant detail. here we have a URL parser and it should properly recognize any URL it is given.

what's not in this class is `next_url()`, which is something I haven't yet found in Ada even. we could scan through a file character by character testing each one, but maybe this is where we can take more liberties in saying, "we want to find _specific kinds of URLs_ and then parse those"